src/core/auth/strings.rs

```diff
@@ -0,0 +1,149 @@
+use bech32::{self, FromBase32, ToBase32, Variant, u5};
+use core::result::Result as result;
+
+fn check_prefix_and_length(prefix: &str, data: &str, length: i32) -> bool {
+    match bech32::decode(data) { 
+        result::Ok(t) => {
+            t.0 == prefix && data.chars().count() == length as usize
+        }
+        result::Err(e) => {
+            eprintln!("{}", e);
+            false
+        },
+    }
+}
+
+// **********  AccAddress *********** //
+pub struct AccAddress {
+    val: String,
+}
+
+fn is_acc_address(data: &str) -> bool {
+    //return check_prefix_and_length("secret", data, 45)
+    check_prefix_and_length("secret", data, 45)
+}
+
+fn to_acc_address(data: ValAddress) -> Option<AccAddress> {
+    let vals = bech32::decode(&data.val);
+    match vals {
+        Ok(t) => {
+            match bech32::encode("secret", t.1, t.2) {
+                Ok(string) => Some(AccAddress {
+                    val: string,
+                }),
+                Err(e) => {
+                    eprintln!("{}", e);
+                    None
+                },
+            }
+        },
+        Err(e) => {
+            eprintln!("{}", e);
+            None
+        },
+    }
+}
+
+
+// **********  ValAddress *********** //
+pub struct ValAddress {
+    val: String,
+}
+
+fn is_val_address(data: &str) -> bool {
+    return check_prefix_and_length("secretvaloper", data, 52);
+}
+
+fn to_val_address(data: AccAddress) -> Option<ValAddress> {
+    let vals = bech32::decode(&data.val);
+    match vals {
+        Ok(t) => {
+            match bech32::encode("secretvaloper", t.1, t.2) {
+                Ok(string) => Some(ValAddress {
+                    val: string,
+                }),
+                Err(e) => {
+                    eprintln!("{}", e);
+                    None
+                },
+            }
+        },
+        Err(e) => {
+            eprintln!("{}", e);
+            None
+        },
+    }
+
+}
+
+// // **********  AccPubKey *********** //
+pub struct AccPubKey {
+    val: String,
+}
+fn is_acc_pubkey(data: &str) -> bool {
+    return check_prefix_and_length("secretpub", data, 77);
+}
+
+fn to_acc_pubkey(data: ValPubKey) -> Option<AccPubKey> {
```

Comment:
All of these options should be replaced with Result<> and in the case of the error, The error should be passed to to the function that's calling. This is done really nicely with the ? operator. On Ok, it will unwrap the value and on Err it will pass the error up to the previous function call

Reply:
https://www.reddit.com/r/rust/comments/a38enj/question_mark_operator/